### ISSUE TITLE
[CDS] Add/port retry_policy to Envoy cluster. 

### ISSUE THREAD

**Issue by stevenzzzz** [envoyproxy/envoy#25751]
*Title*: Support retry_policy at Cluster. 
*Description*: 
> Oftentimes, different clusters have its own retry policies, right now Envoy only support Route based retry policy, which means a user needs to configure every route pointing to that Cluster with exactly the same retry policy. it's an unnecessary waste of RAM considering how "bulky" the retrypolicy proto/impl is. This scenario is oftentimes seen for routes/vhost that points to a external [public] service, e.g. cloud storage. 

I think we should also support RetryPolicy on a cluster, and to support finer grained control, 
make it configurable if a cluster retry policy, when present, should overwrite the route level retry policy, considering that usually route-entry-level retry policy falls back to vhost level retry policy.

**Comment by stevenzzzz** [comment]
@yanavlasov

**Comment by stevenzzzz** [comment]
@htuch @efimki

**Comment by htuch** [comment]
The need for something like this makes sense, but it creates a shared hierarchical policy split across RouteConfiguration (RDS) and Cluster (CDS) consumed by the router. I'm trying to think if we have precedent for this and I'm not sure. One of the advantages of having this policy expressible at cluster level is it becomes available to things like the HTTP/gRPC async clients that only have synthetic route tables. CC @envoyproxy/api-shepherds.

Other related issues:
* https://github.com/envoyproxy/envoy/issues/23060
* https://github.com/envoyproxy/envoy/issues/21131

**Comment by adisuissa** [comment]
I think that having an ability to share cluster-specific policy makes sense, although I'm not certain if there is a precedent or not. As @stevenzzzz said, there should be a clear understanding of config precedence between the different route-level configs and the cluster-level configs.

Alternatives:
* If the main goal is to reduce repeated configs, it may be useful to have a per-route retry-policy _reference-name_ that points to an entry in a map of RetryPolicy configs. This will also allow sharing the same policy across different clusters. The added value is that it reduces the cognitive load when trying to understand a config-dump while debugging an issue. However, this won't support the synthetic route tables case that @htuch pointed out.
* I think this may be achieved by using internal-listeners, although there will probably be some performance implications.

**Comment by stevenzzzz** [comment]
To @adisuissa it's not just about reducing config size. 

Right now the retry_policy is a "backend-oriented" policy, i.e. it only applies to retries to the customer backends. 

But consider this case, in a service callouts through ext_auth or proxy_wasm, those filters can't have their own retries behavior on the grpc services they want to enable different retry polices.

There are quite a few configs that "could either be on cluster, or on route", for example, cluster-specific "custom headers", right now this is only viable for a cluster if we define a weightedclusters in route config. But oftentimes, the public storage services would appreciate some sort of "keys" that's very "cluster-spefic".

**Comment by adisuissa** [comment]
Yes, I agree:
>  However, this won't support the synthetic route tables case that @htuch pointed out.

Making the configuration on the cluster will also provide these capabilities for xDS services that are based on EnvoyGrpc. That would be very useful.

This actually has led me to think whether the policy should be as part of the service-backend config, or as part of the call to the service config.
Consider 2 different routes that use the same external service (e.g., ext_auth). Could there be a different retry-policy configuration depending on the route used, but if the config is only on the cluster-side, then I guess one will need 2 clusters with the different policy.
I personally think setting it on the cluster is better due to simplicity, just pointing out an alternative.

BTW: Looking around the codebase it seems that there are specific filters that define a (different) retry policy for their side-service (e.g., [ALS](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/access_loggers/grpc/v3/als.proto#L99), [gcp_authn](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/gcp_authn/v3/gcp_authn.proto#L31)).

> There are quite a few configs that "could either be on cluster, or on route", for example, cluster-specific "custom headers", right now this is only viable for a cluster if we define a weightedclusters in route config. But oftentimes, the public storage services would appreciate some sort of "keys" that's very "cluster-spefic".

Yes, and we should consider all these cases when deciding the config-precedence between different route-level configs and the cluster-level ones.

**Comment by stevenzzzz** [comment]
I will let you decide. :D
my preference would be adding them incrementally. But if you are proposing some fundamental change, that's fine too.

**Comment by markdroth** [comment]
@htuch 
> The need for something like this makes sense, but it creates a shared hierarchical policy split across RouteConfiguration (RDS) and Cluster (CDS) consumed by the router. I'm trying to think if we have precedent for this and I'm not sure. 

I don't think we have anything like this today, but I agree that it makes sense.  I think one possible approach would be to add a new `typed_per_filter_config` map to `Cluster`, which would fit right into the existing hierarchy of such fields: we already have one in each of [`VirtualHost`](https://github.com/envoyproxy/envoy/blob/6c18a51ea75b8b184e65acdb8364403df4b35e37/api/envoy/config/route/v3/route_components.proto#L168), [`Route`](https://github.com/envoyproxy/envoy/blob/6c18a51ea75b8b184e65acdb8364403df4b35e37/api/envoy/config/route/v3/route_components.proto#L307), and [`WeightedCluster`](https://github.com/envoyproxy/envoy/blob/6c18a51ea75b8b184e65acdb8364403df4b35e37/api/envoy/config/route/v3/route_components.proto#L467), so adding another one in `Cluster` would just fit right in to that established hierarchy.

One issue here is that the retry policy is currently a top-level field in the route rather than being part of the configuration for the router filter.  But that's probably something we should change anyway.  (We've done similar changes before, such as moving the rate limiting config into the filter config.)

> One of the advantages of having this policy expressible at cluster level is it becomes available to things like the HTTP/gRPC async clients that only have synthetic route tables.

I don't know about HTTP clients, but I don't believe that there are any gRPC clients with synthetic route tables.  When we designed xDS support in gRPC, we realized fairly early that we needed to support route configs, because that's where most of the configuration that is interesting to the client is actually located.

@adisuissa 
> Could there be a different retry-policy configuration depending on the route used, but if the config is only on the cluster-side, then I guess one will need 2 clusters with the different policy.

For gRPC, I think it's actually fairly common to apply retry config on a per-RPC-method basis -- some methods make sense to retry, whereas others do not.  I think that in that kind of case, we'll still need to specify the retry config in the routes, not in the cluster.  But that doesn't mean that there won't also be cases where it's useful to configure it in the cluster.

**Comment by stevenzzzz** [comment]
ping @htuch we need someone to commit on this.

**Comment by stevenzzzz** [comment]
@rshriram maybe you want to comment on this feature?

**Comment by htuch** [comment]
I think we should do this, PRs to explore the API as first step welcome. What @markdroth said is a good idea IMHO, and I think we could model this as a per-route override of the HTTP router filter behavior.

**Comment by stevenzzzz** [comment]
heh, let's revive this. 
Could you pls nostale this issue @adisuissa ?

**Comment by adisuissa** [comment]
Reopening (and no-staling) as requested

**Comment by rshriram** [comment]
what is the precedence order here? which one will be preferred if both are specified?

**Comment by stevenzzzz** [comment]
I was thinking the route-level one should take precedence (realized that I said the reverse in this issue). 

OTOH, we should probably have a setting to let users specify the precedence order. 

Looking a little bit deeper into it, it might a a lot of toil for plumbing a route:: space object into the upstream:: spacee, I am not sure how API shepard like this idea. 

Summon @adisuissa

**Comment by adisuissa** [comment]
I agree that route-level should take precedence over the cluster-one (so route overrides virtual-host that overrides cluster).

My initial thought was that a cluster may have a default-retry policy defined, and that will be used in the routing filter.
However, on thinking about it more deeply I think there are 2 issues that may need further consideration:
1) The route and clusters are fetched at different times (non-atomic) which can lead to inconsistent behaviors.
2) When trying to debug why some request behaves in a strange way, looking at the cluster-dump is not very intuitive (but rather at the routing-table).

Taking a step back, is the problem that the config is too verbose and the protos are large, or is it the duplication in the Envoy memory?
If the latter then I think we can use the [ObjectSharedPool](https://github.com/envoyproxy/envoy/blob/b76b08cc56583a09d432ba87094205df9f7e26bc/source/common/shared_pool/shared_pool.h#L35) to reduce the memory usage.

Another question: Can the virtual-host level policy help in this case?

**Comment by htuch** [comment]
> Taking a step back, is the problem that the config is too verbose and the protos are large, or is it the duplication in the Envoy memory?

I think the real issue is that retry is a property of the cluster in many situations, e.g. ext_proc wants this but has no RouteConfiguration. You ned up having to respecify these retry properties again-and-again in every extension like ext_authz that wants this.

FWIW the flip-flop concern is probably there today for per-route filter overrides, since in theory filter chains can update independently of RDS, e.g. via ECDS, probably somewhat tempered by warming.

**Comment by adisuissa** [comment]
> I think the real issue is that retry is a property of the cluster in many situations, e.g. ext_proc wants this but has no RouteConfiguration. You ned up having to respecify these retry properties again-and-again in every extension like ext_authz that wants this.
>

I can see why retry can be considered a property of the cluster, though I assume that it was introduced into the routes as different routes (to the same cluster) may have different retry policy.

It seems that repeated sub-resources can be used in many places, and the issue is not unique to the retry-policy. In that case, I wonder if there could be a single map in a response proto that maps a unique-id to Any config, and fields can point to them. This still requires sending it at least once in a proto response, but allows to keep the same self-contained semantics that responses have today.

> FWIW the flip-flop concern is probably there today for per-route filter overrides, since in theory filter chains can update independently of RDS, e.g. via ECDS, probably somewhat tempered by warming.

I must admit that I haven't thoroughly understood why there was no Route->Cluster warming and why clusters came before listeners (I assume because it was prior to warming), even though I followed some of the issues where this was discussed.

**Comment by markdroth** [comment]
I think the principle we should apply here is that finer-grained settings always override coarser-grained settings, since it's desirable to be able to set coarse-grained defaults and then override them in specific cases as needed.  Applying that principle, I suggest that the precedence order should be:

VirtualHost -> Cluster -> Route -> ClusterWeight

Adi, I don't think we should be specifically concerned here about the fact that the configuration is split across multiple resources and cannot be delivered atomically.  That's a pre-existing problem for xDS (e.g., it also happens for filter configs split across LDS and RDS) and is already something that control planes and data planes need to be able to deal with.

**Comment by stevenzzzz** [comment]
Yeah, @htuch 's case is exactly what I am thinking of: for some "side services", say pointed by ext_proc, or gcp_authn etc. 
The retry-policy from route may not be the one those services need. 

some similar properties (not side-service specific) could be: hedge, tracing, or even per-service header overwrites/customer headers, which I am not chasing after yet. 

(compaction on config side could be another reason). 

totally agreed with @markdroth on that "finer-grained settings always override coarser-grained settings". one thing I am not sure, shall they merge or override. I think overriding is probably cleaner. 

The challenge I am having is that if we do this, API shepards need to agree that api/envoy/config/cluster/** be able to point to api/envoy/config/route/**. (same for code structure). 

On the other hand, I think for ext_proc, we can for now take the easier path to insert a global RetryPolicy into externalProcessor proto, which has several prior art examples in envoy already.

**Comment by markdroth** [comment]
> one thing I am not sure, shall they merge or override. I think overriding is probably cleaner.

It's up to the individual filter to decide how it interprets config overrides at each of these levels.  Most filters choose to use override semantics, because it's so much easier to understand the configuration.  I think we do have a couple of examples of filters that use merge semantics, but I would recommend sticking with override semantics unless there's a strong reason to do otherwise.

> some similar properties (not side-service specific) could be: hedge, tracing, or even per-service header overwrites/customer headers

I think we should ultimately be able to take the same approach for any of those features.

> The challenge I am having is that if we do this, API shepards need to agree that api/envoy/config/cluster/** be able to point to api/envoy/config/route/**.

I don't think we want to do that.  Instead, I think we should do the following:

- Copy the [`RetryPolicy` message](https://github.com/envoyproxy/envoy/blob/12e928ce06a8f0bf32b5e27ec1bda2368cddc920/api/envoy/config/route/v3/route_components.proto#L1390) into the [config for the router filter](https://github.com/envoyproxy/envoy/blob/12e928ce06a8f0bf32b5e27ec1bda2368cddc920/api/envoy/extensions/filters/http/router/v3/router.proto#L27), and add a field of that type in that filter config.
- Deprecate the existing [`VirtualHost.retry_policy`](https://github.com/envoyproxy/envoy/blob/12e928ce06a8f0bf32b5e27ec1bda2368cddc920/api/envoy/config/route/v3/route_components.proto#L191) and [`RouteAction.retry_policy`](https://github.com/envoyproxy/envoy/blob/12e928ce06a8f0bf32b5e27ec1bda2368cddc920/api/envoy/config/route/v3/route_components.proto#L1261) fields.
- Change the implementation of the router filter to prefer using the new field added to its config, but to fall back to using the deprecated fields if the field in its filter config is not set.

**Comment by stevenzzzz** [comment]
> Copy the [RetryPolicy message](https://github.com/envoyproxy/envoy/blob/12e928ce06a8f0bf32b5e27ec1bda2368cddc920/api/envoy/config/route/v3/route_components.proto#L1390) into the [config for the router filter](https://github.com/envoyproxy/envoy/blob/12e928ce06a8f0bf32b5e27ec1bda2368cddc920/api/envoy/extensions/filters/http/router/v3/router.proto#L27), and add a field of that type in that filter config.

Is this idea to plumb retry_policy config in Cluster::typed_per_filter_config with the key "envoy.router"?

I think it's good that we can put everything with a typed_filter_config, but I'm trying understand what the factory look like, does it provide a router::FilterConfig back, from which we can read the retry_policy?

**Comment by markdroth** [comment]
The factory already has a config for the router filter.  What we need to do is to just add a field to that config that contains the retry policy.  And since we don't want the router filter config to depend on the RouteConfig protos, we need to make a copy of the `RetryPolicy` message in the file that contains the config for the router filter.
